lib/action_controller/respond_with.rb

```diff
@@ -34,20 +34,38 @@ module ClassMethods
       #
       # This specifies that the <tt>:create</tt> action and no other responds
       # to <tt>:json</tt>.
+      #
+      # Requests that do not match the mime types defined with <tt>respond_to</tt>
+      # will be executed normally and might raise an
+      # <tt>ActionController::UnknownFormat</tt> if the action invokes the
+      # <tt>respond_with</tt> method. To prevent the action from being executed
+      # use the <tt>:strict</tt> option.
+      #
+      #   respond_to :json, strict: true
+      #
+      # <tt>:strict</tt> will install a <tt>before_action</tt> callback to raise
+      # the <tt>ActionController::UnknownFormat</tt> if the action does not
+      # responds to the request mime type.
       def respond_to(*mimes)
         options = mimes.extract_options!
 
         only_actions   = Array(options.delete(:only)).map(&:to_s)
         except_actions = Array(options.delete(:except)).map(&:to_s)
 
-        new = mimes_for_respond_to.dup
+        hash = mimes_for_respond_to.dup
         mimes.each do |mime|
           mime = mime.to_sym
-          new[mime]          = {}
-          new[mime][:only]   = only_actions   unless only_actions.empty?
-          new[mime][:except] = except_actions unless except_actions.empty?
+          hash[mime]          = {}
+          hash[mime][:only]   = only_actions   unless only_actions.empty?
+          hash[mime][:except] = except_actions unless except_actions.empty?
+        end
+        self.mimes_for_respond_to = hash.freeze
+
+        if options[:strict]
+          prepend_before_action :only => only_actions.presence, :except => except_actions.presence do |controller|
```

Comment:
doing `prepend_before_action :validate_request_format!` was raising some inconsistencies to the test suite, maybe because multiple calls like that was replacing the previous `only`/`except` configuration, so I ended up with this so far, which I not a big fan as the `validate_request_format!` method is currently public.